pkgs/os-specific/linux/plymouth/default.nix

```diff
@@ -1,95 +1,93 @@
 { lib
 , stdenv
-, fetchpatch
 , fetchFromGitLab
+, meson
 , pkg-config
-, autoreconfHook
+, libpng
+, udev
+, pango
+, gtk3
+, libdrm
+, libevdev
+, libxkbcommon
+, xorg
 , libxslt
+, ninja
 , docbook-xsl-nons
-, gettext
-, gtk3
 , systemd
-, pango
-, cairo
-, libdrm
 }:
 
 stdenv.mkDerivation rec {
   pname = "plymouth";
-  version = "unstable-2021-10-18";
-
-  outputs = [
-    "out"
-    "dev"
-  ];
+  version = "unstable-2023-02-22";
 
   src = fetchFromGitLab {
     domain = "gitlab.freedesktop.org";
-    owner = "plymouth";
-    repo = "plymouth";
-    rev = "18363cd887dbfe7e82a2f4cc1a49ef9513919142";
-    sha256 = "sha256-+AP4ALOFdYFt/8MDXjMaHptkogCwK1iXKuza1zfMaws=";
+    owner = pname;
+    repo = pname;
+    rev = "7302b49da8b98d01ab82d23e5be74146065cffd8";
+    hash = "sha256-PgVJezSUd/E3o1YigVto5KmQlflONrmKspOmAHSc5p4=";
   };
 
+  patches = [
+    ./add-option-for-installation-sysconfdir.patch
+    ./dont-create-broken-symlink.patch
+  ];
+
   nativeBuildInputs = [
-    autoreconfHook
-    docbook-xsl-nons
-    gettext
-    libxslt
+    meson
     pkg-config
+    ninja
+    docbook-xsl-nons
   ];
 
   buildInputs = [
-    cairo
+    libpng
+    udev
+    pango
     gtk3
     libdrm
-    pango
+    libevdev
+    libxkbcommon
+    xorg.xkeyboardconfig
+    libxslt
     systemd
   ];
 
-  postPatch = ''
-    sed -i \
-      -e "s#plymouthplugindir=.*#plymouthplugindir=/etc/plymouth/plugins/#" \
-      -e "s#plymouththemedir=.*#plymouththemedir=/etc/plymouth/themes#" \
-      -e "s#plymouthpolicydir=.*#plymouthpolicydir=/etc/plymouth/#" \
-      -e "s#plymouthconfdir=.*#plymouthconfdir=/etc/plymouth/#" \
-      configure.ac
-  '';
-
-  configurePlatforms = [ "host" ];
-
-  configureFlags = [
-    "--enable-documentation"
-    "--enable-drm"
-    "--enable-gtk"
-    "--enable-pango"
-    "--enable-systemd-integration"
-    "--enable-tracing"
-    "--localstatedir=/var"
+  mesonFlags = [
     "--sysconfdir=/etc"
```

Comment:
That is IMO more fragile than patch. New compile definitions can be introduced and we will not know about that until Plymouth breaks at runtime.

Reply:
Yes but there's too much inconsistent logic in the `meson.build` files that will probably not install right with this (e.g. they could add more files and then those would have to be patched out too)

They're much less likely to add new config dirs that need to be changed at runtime than adding new files to install with inconsistent meson rules (and previously it was just these four that we manually patched with autotools too)

So idk if I'd call it more fragile than the patch, but admittedly a patch to fix their inconsistent logic is something potentially upstreamable

Reply:
(Also I did originally start by building off the patch to fix more meson logic, but it would have been either quite a lot just to get the compile definitions to come out as we need or I would have had to straight up add new options to allow overriding them separately from the logic)

Reply:
By fragile I meant that failures would not be revealed until runtime. By having to explicitly fix installation directories, we would know whenever a new file is added since the build would fail.

Do you have a list of problems you have found and tried to fix? It might be a good idea to record them as a Nixpkgs or upstream issue if we cannot resolve them now.

Reply:
> By fragile I meant that failures would not be revealed until runtime. By having to explicitly fix installation directories, we would know whenever a new file is added since the build would fail.

It already was silently failing at runtime as-is (took me a while to figure that out poking around in a minimal initrd environment...) but I do get your point and it is valid

> Do you have a list of problems you have found and tried to fix? It might be a good idea to record them as a Nixpkgs or upstream issue if we cannot resolve them now.

Yep! I should still have the wip fixes in my local plymouth clone so I'll grab those and make a list of the over-arching problems and post them here later

The biggest is that the meson fes seem to assume that where it installs the files to will be the place it will find them at runtime, which is very untrue in our case and especially around plugins and stuff that it needs to look in `/etc` for but install to `$out/lib` iirc

Reply:
The more I look, there's really so many places where they are installed somewhere based on the variables that go into those compile definitions. Each plugin uses `plymouth_plugin_path` to install, themes are using `plugin_themes_path` to install too, etc

I keep trying to minimize and find the least invasive changes (that would still make sense for upstream) on the meson files to get them to install stuff to the normal place but allow us to override the compile definitions that go in for looking at things at runtime, but I really can't find a way that would be reasonable to upstream

Perhaps we can just override CFLAGS to get our definitions later in the command arguments when compiling (so that way what those variables are will not matter) or maybe upstream would go for just adding options to set those four dirs just for the compile definitions?

I'm not sure what's best, as I'm really not that good with meson stuff and also don't know very well what I'm doing

Do you have any other ideas @jtojnar?

Reply:
Overriding CFLAGS sounds like more space for silent breakage. If it is that many paths, maybe doing the `DESTDIR` hack as mentioned in https://github.com/NixOS/nixpkgs/pull/217728#discussion_r1114844731 would be cleaner.

Reply:
> Overriding CFLAGS sounds like more space for silent breakage.

That was my thought on it too

> If it is that many paths, maybe doing the `DESTDIR` hack as mentioned in [#217728 (comment)](https://github.com/NixOS/nixpkgs/pull/217728#discussion_r1114844731) would be cleaner.

Oh I missed that comment! I'll try to hack on that today or tomorrow to gauge feasibility

Thanks!

Reply:
Gentle ping on the above question/diff if you have any time @jtojnar

Thank you again for your feedback!